plugins/module_utils/network/ios/facts/vlans/vlans.py

```diff
@@ -44,161 +45,67 @@ def __init__(self, module, subspec="config", options="options"):
-                        config["state"] = "active"
-                    config["shutdown"] = "enabled"
-                else:
-                    if conf[state_idx] == "suspended":
-                        config["state"] = "suspend"
-                    elif conf[state_idx] == "active":
-                        config["state"] = "active"
-                    config["shutdown"] = "disabled"
-            except IndexError:
-                pass
-        elif vlan_info == "Type" and "VLAN Type" not in conf:
-            conf = list(filter(None, conf.split(" ")))
-            config["mtu"] = int(conf[3])
-        elif vlan_info == "Remote":
-            if len(conf.split(",")) > 1 or conf.isdigit():
-                remote_span_vlan = []
-                if len(conf.split(",")) > 1:
-                    remote_span_vlan = conf.split(",")
-                else:
-                    remote_span_vlan.append(conf)
-                remote_span = []
-                for each in remote_span_vlan:
-                    split_sp_list = each.split("-")
-                    if len(split_sp_list) > 1:  # break range
-                        for r_sp in range(int(split_sp_list[0]), int(split_sp_list[1]) + 1):
-                            remote_span.append(r_sp)
-                    else:
-                        remote_span.append(int(each))
-                config["remote_span"] = remote_span
-
-        elif vlan_info == "Private" and "Primary Secondary" not in conf:
-            conf = list(filter(None, conf.split(" ")))
-
-            pri_idx = 0
-            sec_idx = 1
-            priv_type_idx = 2
-
-            config["tmp_pvlans"] = {
-                "primary": conf[pri_idx],
-                "secondary": conf[sec_idx],
-                "sec_type": conf[priv_type_idx],
-            }
-        return utils.remove_empties(config)
-
-    def parse_vlan_config(self, vlan_conf):
-        vlan_list = list()
-
-        re1 = re.compile(r"^vlan configuration +(?P<vlan>\d+)$")
-        re2 = re.compile(r"^member +(evpn\-instance +(?P<evi>\d+) )?vni (?P<vni>[\d\-]+)$")
 
-        for line in vlan_conf.splitlines():
-            line = line.strip()
-
-            m = re1.match(line)
-            if m:
-                vlan = m.groupdict()["vlan"]
-                vlan_dict = {"vlan_id": vlan}
-                continue
-
-            m = re2.match(line)
-            if m:
-                group = m.groupdict()
-                vlan_dict.update({"member": {}})
-                vlan_dict["member"].update({"vni": group["vni"]})
-                if group["evi"]:
-                    vlan_dict["member"].update({"evi": group["evi"]})
-                vlan_list.append(vlan_dict)
-
-        return vlan_list
-
-    def parse_vlan(self, data):
         objs = []
         mtu_objs = []
         remote_objs = []
         final_objs = []
         pvlan_objs = []
 
+        if not data:
+            data = self.get_vlans_data(connection)
+
```

Comment:
Wouldn't it be simpler to move line #100 to line #98, i.e., connect to the device and fetch both the configuration stanzas when "data" is unavailable, a.k.a., `state != parsed`.

Then `populate_vlans_config_facts()` can just deal with parsing the `conf_data`.

Reply:
Yeah, That is doable but in that case, we forcefully feed vlans and vlan configuration data to the parsers, which makes the parsers hit vlans-specific lines that are not being processed by parsers, so to handle that it's separated. Yeah, for parsed it would pass two `if not data:` conditions one for vlans which is done the older way, and the other for vlan configuration.